pkg/operator/ceph/controllers/nodedrain/reconcile.go

```diff
@@ -0,0 +1,143 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "nodedrain-controller")
+	// Implement reconcile.Reconciler so the controller can reconcile objects
+	_ reconcile.Reconciler = &ReconcileNode{}
+)
+
+const (
+	nodeHostNameKey = "kubernetes.io/hostname"
+)
+
+// ReconcileNode reconciles ReplicaSets
+type ReconcileNode struct {
+	// client can be used to retrieve objects from the APIServer.
+	scheme  *runtime.Scheme
+	client  client.Client
+	options *controllerconfig.Options
+}
+
+// Reconcile reconciles a node and ensures that it has a drain-detection deployment
+// attached to it.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNode) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// workaround because the rook logging mechanism is not compatible with the controller-runtime loggin interface
+	result, err := r.reconcile(request)
+	if err != nil {
+		logger.Error(err)
+	}
+	return result, err
+}
+
+func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result, error) {
+
+	logger.Debugf("reconciling node: %s", request.Name)
+
+	node := &corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: request.Name}}
+
+	err := r.client.Get(context.TODO(), request.NamespacedName, node)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Could not get node %s", request.NamespacedName)
+	}
+
+	nodeHostnameLabel, ok := node.ObjectMeta.Labels[nodeHostNameKey]
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("Label key %s does not exist on node %s", nodeHostNameKey, request.NamespacedName)
+	}
+
+	// Create or Update the deployment default/foo
+	deploy := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("rook-ceph-canary-%s", request.Name),
+			Namespace: r.options.OperatorNamespace,
+		},
+	}
+
+	// CreateOrUpdate the deployment
+	mutateFunc := func() error {
+
+		// lablels for the pod, the deployment, and the deploymentSelector
+		deploymentLabels := map[string]string{
+			nodeHostNameKey: nodeHostnameLabel,
+			k8sutil.AppAttr: CanaryAppName,
+		}
+
+		nodeSelector := map[string]string{nodeHostNameKey: nodeHostnameLabel}
+
+		// Deployment selector is immutable so we set this value only if
+		// a new object is going to be created
+		if deploy.ObjectMeta.CreationTimestamp.IsZero() {
+			deploy.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: deploymentLabels,
+			}
+		}
+
+		// update
+		deploy.ObjectMeta.Labels = deploymentLabels
+
+		// update the Deployment pod template
+		deploy.Spec.Template = corev1.PodTemplateSpec{
```

Comment:
+1

Reply:
It is possible to know which nodes have OSDs. We'd have to monitor all OSD pods and schedule a canary on whichever nodes they're scheduled to. I was worried about that being inaccurate, but that's another scenario where a delay in information getting to us is not dangerous.

If possible, I'd like to do that in another PR. What we have currently doesn't behave too differently from a Daemonset. I have a lot of optimizations and caching that I'd like to add, but I think it would make more sense to combine that effort with the integration with the fencing logic.

I forsee some future PRs wrt this:
- @ashishranjan738 's fencing work which is blocked by this merging
- A PR that integrates the information discovery between the two and lazy loads the information when it is accessed, caches it, and then only reloads it after an expiry.
- A more robust abstraction on orthogonal reconciliation procedures (mon PDB, osd PDB) within the same Reconciler.